### ISSUE TITLE
Add information about current best point in callback of `optimize.minimize` 

### ISSUE THREAD

**Issue by schmoelder** [scipy/scipy#21061]
When specifying a callback function, it's possible to specify a callable with the following signature ([see documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html))

```
callback(intermediate_result: OptimizeResult)
```

In the documentation, it states:

> [...] where intermediate_result is a keyword parameter containing an [OptimizeResult](https://docs.scipy.org/doc/scipy-1.14.0/reference/generated/scipy.optimize.OptimizeResult.html#scipy.optimize.OptimizeResult) with attributes `x` and `fun`, the present values of the parameter vector and objective function.

While it's good to know the current state, I was wondering if it would also be possible to provide information about the current best point, either by adding this to the intermediate result itself, or by adding a new optional argument in the signature.

**Comment by schmoelder** [comment]
Sorry, I just found that this is a duplicate of several other issues. Still, would be a nice feature! :nerd_face:

**Comment by andyfaff** [comment]
> , I was wondering if it would also be possible to provide information about the current best point,

it provides exactly that.

**Comment by schmoelder** [comment]
Hi Andy,

thanks for your reply!

> > , I was wondering if it would also be possible to provide information about the current best point,
> 
> it provides exactly that.

This is also how I had understood it but some tests were not consistent which is why assumed it wasn't. I ran some more tests and it turns out it was an issue with the new `COBYQA` optimizer that does in fact simply use the latest point. I will contact the devs and open an issue.

Note, the documentation could be more a bit more explicit here. As mentioned above, it currently states 

> the present values of the parameter vector and objective function.

So maybe something like the following could clear this up:

> the current *best* values of the parameter vector and objective function.

**Comment by mdhaber** [comment]
The callback is intended to give information about the *current* point because that provides more information than the best point ever visited. If you need the keep track of the best point visited, you do that with simple logic in the callback function.

**Comment by schmoelder** [comment]
Hi Matt,

thanks for your insight! I understand your reasoning and I'm fine with it any way. I might have misunderstood @andyfaff's comment.

Note, it appears that in some methods, e.g. `NelderMead`, some sorting seems to be performed (see [here](https://github.com/scipy/scipy/blob/3959e8b28a8fdf9f6803fe34bf8f0e34909d840d/scipy/optimize/_optimize.py#L898)). But I don't think this is a bug, it's more to do with how the algorithm works, I suppose.

---

> ```
> if intermediate_result.fun < incumbent_fun
>     incumbent_fun = intermediate_result.fun
>     incumbent_x = intermediate_result.x
> ```

I'm wondering, how would this work when constraints are involved? Suppose in a previous iteration, some constraints were violated but the function value was low. Now the function value might be "worse" but constraints are met (which is preferable).

**Comment by mdhaber** [comment]
> Suppose in a previous iteration, some constraints were violated but the function value was low.

Yeah, that's why constraint information should be added to the callback. The condition would become like:

```python3
if intermediate_result.fun < incumbent_fun and <constraints are satisfied>
```

I had written in gh-17486 that "The intent here is just to pave the way" by allowing an `OptimizeResult` rather than a single array to be passed to the objective function. I had hoped that other contributors/reviewers would populate the `OptimizeResult` with the desired information. It's quite easy to do; it just takes a bit of patience to find out which variables in the code are the correct ones to include, write the tests to confirm that everything is working as expected, and document what is being added. I've have several [other](https://github.com/scipy/scipy/pulls/mdhaber) SciPy things going on since then, so I had to prioritize my time.

> But I don't think this is a bug, it's more to do with how the algorithm works, I suppose.

Hmm in gh-17486 I included in the `OptimizeResult` whatever was being passed directly to the callback before. The documentation stated that `xk` was "the _current_ parameter vector", so I assumed that was interpreted consistently for all methods. It's possible that when the callback function was originally added, the meaning of `xk` was not interpreted the same way as for other methods, in which case that might be considered a bug. Yes, one would need to look at how the algorithm is working to see whether that is the "current" point or "best" point for `NelderMead`.

**Comment by andyfaff** [comment]
I have definitely been working on the basis of `intermediate_result.x` to be the best result so far. That's because  for most of the optimisers it is, I think. It's more useful than the alternative (one could just track that in the objective function).

Take differential evolution. There multiple evaluations are used per iteration. We don't want to use a callback multiple times per iteration, it wouldn't make sense. Because it would get called way too many times, and the fitness would be oscillating widely. Plus the overhead of calling the callback would be too high.

Whilst that's a case at one end of the spectrum, the arguments still apply across it. I would expect callback to be used at the completion of an outer loop with the best result so far.

**Comment by andyfaff** [comment]
If cobyqa is not returning best so far, then that should be amended.

**Comment by mdhaber** [comment]
Ah, well there we have it!

I don't think I changed anything in the `intermediate_result` vs what was being passed to the callback before, so I don't think I've changed anything based on my interpretation of "where ``xk`` is the current parameter vector". 

But if two maintainers had different ideas of what that means, I can imagine that the code might be inconsistent, too.

> It's more useful than the alternative (one could just track that in the objective function)

But often one would have to filter out the actual point being visited vs what evaluations are being done to estimate derivatives. I thought *that* was the advantage of the `callback`. I think it's easier for a user to remember the best value seen so far (or whatever they want to remember) than to filter out the calls to the objective function that they are not interested in.

Feel free to go in whichever direction, though. I was just answering from what I thought I knew - and I thought I knew, having written gh-17486!

**Comment by andyfaff** [comment]
I suppose it's a subtlety. In the context of all the solvers when the callbacks are called, "where xk is the current parameter vector" could be rewritten "where xk is the best parameter vector".

**Comment by mdhaber** [comment]
If that's actually what they do, I suppose so. But I think since that also is strictly *less* information that returning the parameter vector of the current iteration, it also restricts what heuristics one can run to decide whether to terminate early or not.

**Comment by mdhaber** [comment]
OK, I performed my own checks by forcibly lifting the objective function after each call to the callback, and I agree that most optimizers stubbornly refuse to provide `callback` with information about the points they are actually visiting : ) Some of them seem to do this by quitting if the objective function value increases, citing "loss of precision". `dual_annealing` and `direct` seems to refuse to end an "iteration" until they find a lower value of the objective function. COBYLA, COBYQA, SLSQP, SHGO, and `basinhopping` seem to be exceptions, not hesitating to report a higher value of the objective function to the callback. `powell` and `l-bfgs-b` seem to avoid reporting a higher objective function value during iterations. However, they do seem to re-evaluate the objective function before the last call to the callback, and they report it honestly even if the objective function has been increased since the previous call.

<details>

```python3
import numpy as np
from scipy import optimize

rng = np.random.default_rng(348923598345)

fvals = []
fvals_real = []

# objective function constant increases with each call to callback
i = [0]
i_inc = 100

def f(x):
    res = optimize.rosen(x) + i[0]
    fvals_real.append(res)
    return res

def callback(x):    
    fvals.append(f(x))
    i[0] += i_inc

# res = optimize.shgo(f, optimize.Bounds([-10]*10, [10]*10), callback=callback)

method = 'l-bfgs-b'
res = optimize.minimize(f, (rng.random(10)-0.5)*5,
                        jac=optimize.rosen_der, hess=optimize.rosen_hess, 
                        method=method, callback=callback)
print(fvals)
print(fvals_real)
```

</details>

**Comment by schmoelder** [comment]
> If cobyqa is not returning best so far, then that should be amended.

Currently, they use the latest point, but I added a PR to return the current best point. However, I think it would be best to come to consensus first about what is actually expected / consistent.

---

I fully understand both perspectives on what would be desired. I personally mostly use population based algorithms so `xk` being the last value does not really make sens to me since it would just be one random individual of the many evaluated in the current iteration.

Moreover, one of the features of using the callback is to raise a `StopIteration`. To use this properly, I would expect to evaluate the current best point and not the latest one. To me, the callback conceptually is called at the end of an outer loop; it's the intermediate results that would have been returned if e.g. `maxiter` had been reached.

---

I had also performed some more checks, on the behavior of different methods, mostly from the `minimize`, don't know if they are useful to anyone moving forward:

<details><summary>Details</summary>
<p>

```Python

import hashlib
import warnings

import numpy as np
from scipy.optimize import (
    minimize, OptimizeResult, OptimizeWarning, rosen, rosen_der, rosen_hess
)


# %%
def random_objective(x):
    h = hashlib.sha256(x.tobytes()).digest()
    hash_int = int.from_bytes(h, byteorder='big')

    # Normalize the integer to a float between 0 and 1
    max_int = 2**256 - 1
    f = hash_int / max_int

    return f


def random_objective_jac(x):
    df_dx = np.random.random()
    return df_dx


def random_objective_hess(x):
    d2f_dxdx = np.random.random()
    return d2f_dxdx


def square(x):
    f = x**2
    return f


def square_jac(x):
    df_dx = 2*x
    return df_dx


def square_hess(x):
    d2f_dxdx = 2
    return d2f_dxdx



# %%
class TestCallback():
    def __init__(
            self,
            method,
            objective='random',
            callback_signature='intermediate_result',
            verbose=True,
            print_results=True
            ):
        self.method = method
        self.callback_signature = callback_signature

        self.verbose = verbose
        self.print_results = print_results

        self.jac = None
        self.hess = None
        if objective == 'random':
            self.x0 = np.array([10.0])
            self._objective = random_objective
            self.jac = random_objective_jac
            self.hess = random_objective_hess
        elif objective == 'square':
            self.x0 = np.array([10.0])
            self._objective = square
            self.jac = square_jac
            self.hess = square_hess
        elif objective == 'rosenbrock':
            self.x0 = np.array([10.0, 10.0])
            self._objective = rosen
            self.jac = rosen_der
            self.hess = rosen_hess
        else:
            raise ValueError("Unknown objective")

        if callback_signature == 'xk':
            self.callback = self.xk
        elif callback_signature == 'intermediate_result':
            self.callback = self.intermediate_result
        else:
            raise ValueError("Unknown callback")

        self.f_min = np.inf
        self.f_is_f_min_counter = 0
        self.f_not_f_min_counter = 0
        self.f_is_f_last_counter = 0
        self.f_not_f_last_counter = 0

        self.x_min = np.inf
        self.x_is_x_min_counter = 0
        self.x_not_x_min_counter = 0
        self.x_is_x_last_counter = 0
        self.x_not_x_last_counter = 0

    def objective(self, x, store_as_last=True):
        f = self._objective(x)

        if store_as_last:
            self.x_last = x
            self.f_last = f

        return f

    def _callback(self, x, f):
        self.x_is_x_min = np.allclose(x, self.x_min)
        if self.x_is_x_min:
            self.x_is_x_min_counter += 1
        else:
            self.x_not_x_min_counter += 1

        self.x_is_x_last = np.allclose(x, self.x_last)
        if self.x_is_x_last:
            self.x_is_x_last_counter += 1
        else:
            self.x_not_x_last_counter += 1

        self.f_is_f_min = np.allclose(f, self.f_min)
        if self.f_is_f_min:
            self.f_is_f_min_counter += 1
        else:
            self.f_not_f_min_counter += 1

        self.f_is_f_last = np.allclose(f, self.f_last)
        if self.f_is_f_last:
            self.f_is_f_last_counter += 1
        else:
            self.f_not_f_last_counter += 1

        if self.verbose:
            print("Evaluating Callback")
            print(f"x: {x}, x_min: {self.x_min}")
            print(f"x is x_min: {self.x_is_x_min}")
            print(f"x is x_last: {self.x_is_x_last}")
            print(f"f: {f}, f_min: {self.f_min}")
            print(f"f is f_min: {self.f_is_f_min}")
            print(f"f is f_last: {self.f_is_f_last}")

    def xk(self, xk: np.ndarray):
        x = xk
        f = self.objective(x, store_as_last=False)

        self._callback(x, f)

    def intermediate_result(self, intermediate_result: OptimizeResult):
        x = intermediate_result.x
        f = intermediate_result.fun

        self._callback(x, f)

    def run(self):
        with warnings.catch_warnings():
            warnings.filterwarnings('ignore', category=OptimizeWarning)
            warnings.filterwarnings('ignore', category=RuntimeWarning)

            result = minimize(
                self.objective,
                self.x0,
                method=self.method,
                callback=self.callback,
                jac=self.jac,
                hess=self.hess,
            )

        if self.print_results:
            print(result)


# %%

verbose = True
verbose = False
print_results = False
objective = 'random'
objective = 'square'
objective = 'rosenbrock'

methods_xk = [
    'Nelder-Mead',
    'Powell',
    'cg',
    'BFGS',
    'Newton-CG',
    'L-BFGS-B',
    'TNC',
    'cobyla',
    'cobyqa',
    'SLSQP',
    # 'trust-constr',
    'dogleg',
    'trust-ncg',
    'trust-exact',
    'trust-krylov',
]

tests_xk = [
    TestCallback(method, objective, 'xk', verbose, print_results)
    for method in methods_xk
]


methods_intermediate = [
    'Nelder-Mead',
    'Powell',
    'cg',
    'BFGS',
    'Newton-CG',
    'L-BFGS-B',
    # 'TNC',
    # 'cobyla',
    'cobyqa',
    # 'SLSQP',
    'trust-constr',
    'dogleg',
    'trust-ncg',
    'trust-exact',
    'trust-krylov',
]


tests_intermediate = [
    TestCallback(method, objective, 'intermediate_result', verbose, print_results)
    for method in methods_intermediate
]

for test in tests_xk + tests_intermediate:
    print(test.callback_signature)
    print(test.method)
    test.run()
    print(f"- x_is_x_min_counter: {test.x_is_x_min_counter}")
    print(f"- x_not_x_min_counter: {test.x_not_x_min_counter}")
    print(f"- x_is_x_last_counter: {test.x_is_x_last_counter}")
    print(f"- x_not_x_last_counter: {test.x_not_x_last_counter}")
    print(f"- f_is_f_min_counter: {test.f_is_f_min_counter}")
    print(f"- f_not_f_min_counter: {test.f_not_f_min_counter}")
    print(f"- f_is_f_last_counter: {test.f_is_f_last_counter}")
    print(f"- f_not_f_last_counter: {test.f_not_f_last_counter}")

```

</p>
</details>

**Comment by andyfaff** [comment]
> To me, the callback conceptually is called at the end of an outer loop;

That is correct callbacks are called once per outer loop (at the start or end depending on the logic). This may not be the case for the recently added cobyqa.

> it's the intermediate results that would have been returned if e.g. maxiter had been reached.

I prefer this behaviour, is what I've been working to, and what I'd like new solvers to work to. Checking that the callback possesses a monotonically decreasing `intermediate_result.fun` is a condition for this (I'll call this behaviour a monotonic solver).

On a `rosen` 3D test problem:

- `direct`, `shgo`, `cobyla` don't follow this behaviour.
- `slsqp` is sometimes monotonic depending on the starting point.
- `nelder-mead`, `bfgs`, `l-bfgs-b`, `powell`, `trust-constr`, `dual_annealing`, `differential_evolution` do obey this behaviour.
 
I didn't look at others. Over time it may be possible for a transition to the latter list.

> To me, this seemed to unambiguously refer to the parameter vector of the present iteration

Unfortunately that is no help. What does the "parameter vector of the present iteration" represent for each of the non monotonic solvers? It's not outlined in any of the docstrings, and it's certainly not always the vector that would be returned if the minimiser were to stop at that point. Unless the user possesses knowledge of the solver internals it would be unwise to `StopIteration` based on the callback parameter vector.

In contrast, it's clearly understood for a monotonic solver what  `intermediate_result.fun`/`intermediate_result.x` are. I propose it's a sensible default for all the different types of algorithms. If a solver wants to report other things then they can add other object attributes.

Solvers that aren't monotonic don't always have the same reason for being that way. It's hard to wade through cobyla and slsqp because they're in Fortran. It's hard to go through shgo and direct because their code is complex.

An aside:
depending on the minimiser the objective function will visit higher energy points in between callbacks. It has to by definition for numerical differentiation, it's almost certainly True for global minimisers, and will happen for derivative free minimisers like nelder-mead.
It's also possible that the objective function visits a lower energy than that reported at the next callback, this is sometimes experienced during a line-search. A larger step could be investigated first (with a lower energy), but a smaller step size might be more optimal (with a slightly higher energy). After taking that smaller step the termination criteria may be met, and the solver stops. Then the final energy won't be the smallest energy visited, but that final energy should be equal to that of the last callback.

**Comment by mdhaber** [comment]
> Unfortunately that is no help.

I hope you mean the docstring?

**Comment by andyfaff** [comment]
The `minimize` docstring is says "present value of the parameter vector" and "current parameter vector". On the face of it the statement is unambiguous. What "current parameter vector" means in practice is unclear without knowing the internals of how a solver works.
 
For example, in the SLSQP code the callback is called [once per (major) iteration](https://github.com/scipy/scipy/blob/f46da3c3e380862b0a1dfae75ffd9afa9c6890f7/scipy/optimize/_slsqp_py.py#L443) of an outer loop. The SLSQP fortran comment [says](https://github.com/scipy/scipy/blob/f46da3c3e380862b0a1dfae75ffd9afa9c6890f7/scipy/optimize/slsqp/slsqp_optmz.f#L77):

```fortran
C*  * X()            X() STORES THE CURRENT ITERATE OF THE N VECTOR X  *
C*                   ON ENTRY X() MUST BE INITIALIZED. ON EXIT X()     *
C*                   STORES THE SOLUTION VECTOR X IF MODE = 0.         *
```

This 'current iterate' value of X is provided to the callback after the increase in major iteration is detected. If MODE is not 0 the solve is still in progress. However, X is not always the best solution so far (callback indicates non-monotonic `result.fun`), so what does X represent in terms of solver state (besides an increment in major iteration)?

Or take `optimize.direct`; at the end of a solve the number of callbacks is the same as the number of iterations (once per major iteration loop), but the number of function evaluations is much larger than the number of callbacks, and the callback has non-monotonic `f(X)`. So for each callback what does X actually represent in terms of solver state?

@czgdp1807, @Stefan-Endres, for direct and shgo, can you furnish further information on what `xk` represents each time a `callback` is called?

**Comment by andyfaff** [comment]
Ideally we could learn from the `basinhopping` docstring, where it's reasonably clear what each callback is given:

> A callback function which will be called for all minima found. ``x`` and ``f`` are the coordinates and function value of the trial minimum, and ``accept`` is whether that minimum was accepted. This can be used, for example, to save the lowest N minima found.

For `basinhopping` each iteration represents one step with a local minimisation at that step. Given that steps to a higher energy are possible via a Monte Carlo acceptance `callback(xk)` will not be monotonically decreasing. Internally the solver does keep track of the lowest energy visited for the final result.
